applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_solid_element.cpp

```diff
@@ -0,0 +1,321 @@
+        for (unsigned k = 0; k < dimension; ++k)
+            rValues[index + k] = adjoint_displacement[k];
+    }
+    KRATOS_CATCH("");
+}
+
+template <class TPrimalElement>
+void AdjointSolidElement<TPrimalElement>::EquationIdVector(EquationIdVectorType& rResult,
+                                                           ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY;
+    auto& r_geom = mPrimalElement.GetGeometry();
+    const unsigned number_of_nodes = r_geom.size();
+    const unsigned dimension = r_geom.WorkingSpaceDimension();
+
+    if (rResult.size() != dimension * number_of_nodes)
+        rResult.resize(dimension * number_of_nodes, false);
+
+    const unsigned pos = r_geom[0].GetDofPosition(ADJOINT_DISPLACEMENT_X);
+
+    if (dimension == 2)
+    {
+        for (unsigned i = 0; i < number_of_nodes; ++i)
+        {
+            const unsigned index = i * 2;
+            rResult[index] = r_geom[i].GetDof(ADJOINT_DISPLACEMENT_X, pos).EquationId();
+            rResult[index + 1] =
+                r_geom[i].GetDof(ADJOINT_DISPLACEMENT_Y, pos + 1).EquationId();
+        }
+    }
+    else
+    {
+        for (unsigned i = 0; i < number_of_nodes; ++i)
+        {
+            const unsigned index = i * 3;
+            rResult[index] = r_geom[i].GetDof(ADJOINT_DISPLACEMENT_X, pos).EquationId();
+            rResult[index + 1] =
+                r_geom[i].GetDof(ADJOINT_DISPLACEMENT_Y, pos + 1).EquationId();
+            rResult[index + 2] =
+                r_geom[i].GetDof(ADJOINT_DISPLACEMENT_Z, pos + 2).EquationId();
+        }
+    }
+    KRATOS_CATCH("");
+}
+
+template <class TPrimalElement>
+void AdjointSolidElement<TPrimalElement>::GetDofList(DofsVectorType& rElementalDofList,
+                                                     ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY;
+    auto& r_geom = mPrimalElement.GetGeometry();
+    const unsigned number_of_nodes = r_geom.size();
+    const unsigned dimension = r_geom.WorkingSpaceDimension();
+    rElementalDofList.resize(0);
+    rElementalDofList.reserve(dimension * number_of_nodes);
+
+    if (dimension == 2)
+    {
+        for (unsigned i = 0; i < number_of_nodes; ++i)
+        {
+            rElementalDofList.push_back(r_geom[i].pGetDof(ADJOINT_DISPLACEMENT_X));
+            rElementalDofList.push_back(r_geom[i].pGetDof(ADJOINT_DISPLACEMENT_Y));
+        }
+    }
+    else
+    {
+        for (unsigned int i = 0; i < number_of_nodes; ++i)
+        {
+            rElementalDofList.push_back(r_geom[i].pGetDof(ADJOINT_DISPLACEMENT_X));
+            rElementalDofList.push_back(r_geom[i].pGetDof(ADJOINT_DISPLACEMENT_Y));
+            rElementalDofList.push_back(r_geom[i].pGetDof(ADJOINT_DISPLACEMENT_Z));
+        }
+    }
+    KRATOS_CATCH("");
+}
+
+template <class TPrimalElement>
+int AdjointSolidElement<TPrimalElement>::Check(const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY;
+    return mPrimalElement.Check(rCurrentProcessInfo);
```

Comment:
Maybe you check here additionally for `ADJOINT_DISPLACEMENT`.

Reply:
I wonder how calling the `mPrimalElement.Check()` does not give an error for you, because there the dofs are checked `KRATOS_CHECK_DOF_IN_NODE(DISPLACEMENT_X, rnode)`. In our finite differencing elements this was always an issue and we ended up copying the check function... (ugly)